text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,121 @@
+
+- **Create Tenant Collection Topics**: This approach involves creating another another set of topics similar to any other tenant. These new set of topics will store messages from all tenants in the FOLIO cluster and will be named in a similar scheme as regular tenant topics are named. The name of this tenant collection will match [A-Z][A-Z0-9]+. This will preserve Kafka subscription patterns during consumption. This approach has been selected for this RFC.
+
+
+## Detailed Explanation/Design
+To promote flexibility of FOLIO implementation patterns, the switch to single topics will be configurable via environment variables. The current naming scheme for topics created by folio-kafka-wrapper is like so:
+>[environment].[namespace].[tenant].[eventtype]
+
+An example: `folio.Default.diku.DI_COMPLETED`. The naming structure is preserved by declaring one more "tenant" where messages for all tenants on the cluster is received for an application concern. e.g. `folio.Default.ALL.DI_COMPLETED`.<p>
+This will allow subscription definitions to remain the same. it will also allow steady migration to single topics for each module without worrying if a consuming module is listening to a single topic as well.
+### Module Configuration
+[folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper) will check for the existence of an environment variable called `KAFKA_PRODUCER_TENANT_COLLECTION`. With this variable unset, the module will produce messages to topics for each tenant. if the variable is set to a value like `ALL`, the module will produce to one topic belonging to the "ALL tenant collection" e.g. `folio.Default.ALL.DI_COMPLETED`.<p>
+In an ideal scenario, the value of the environment variable would be the same for all modules that need to make the switch to single topics.
+### Consuming
+In [folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper), a sample of a subscription pattern is defined below
+
+>folio\.Default\.\w{1,}\.DI_COMPLETED
+
+where `folio` is the cluster name, `Default` is the namespace then a wildcard to capture tenant ids followed by the event type(`DI_COMPLETED`) that the topic is supposed to contain.
+Subscriptions are created using this interface:
+```
+public static SubscriptionDefinition createSubscriptionDefinition(String env, String nameSpace, String eventType)
+```
+With the subscription pattern, consumers will be able to consume from tenant topics and single topics alike. Safeguards can be added to ensure that a message's ownership can be determined. These include:
+- Enriching messages produced by folio-kafka-wrapper with a tenant id before the message is sent.
+- Highlighting messages consumed that don't have a tenant id. Orphan messages can be logged partially or with the full contents. They can also be placed in a dead queue/topic for further review.
```

Comment:
To my knowledge, the tenant id is not derived from the topic name. It is derived from internal data structures inside the message.
In addition, this will be a soft migration. I wrote earlier, that these safeguards CAN be added. Enforcing tenant Id in the header can happen later. This will allow migrations to be easier.